File: 0199-binary-tree-right-side-view/0199-binary-tree-right-side-view.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var rightSideView = function (root) {
  const res = [];
  if (!root) {
    return res;
  }
  const queue = [];
  queue.push(root);

  while (queue.length > 0) {
    const newLevel = [];
    const n = queue.length;

    for (let i = 0; i < n; i++) {
      const node = queue.shift();
      if (node) {
        newLevel.push(node.val);
        if (node.left) {
          queue.push(node.left);
        }
        if (node.right) {
          queue.push(node.right);
        }
      }
    }
    res.push(newLevel[newLevel.length - 1]);
  }
  return res;
};

File: 0199-binary-tree-right-side-view/0199-binary-tree-right-side-view.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    vector<int> rightSideView(TreeNode *root)
    {
        vector<int> res{};
        if (root == nullptr)
        {
            return res;
        }
        queue<TreeNode *> q;
        q.push(root);

        while (q.size() > 0)
        {
            vector<int> newLevel{};

            int n = q.size();

            for (int i = 0; i < n; i++)
            {
                TreeNode *node = q.front();
                newLevel.push_back(node->val);
                q.pop();
                if (node->left)
                {
                    q.push(node->left);
                }
                if (node->right)
                {
                    q.push(node->right);
                }
            }
            res.push_back(newLevel[newLevel.size() - 1]);
        }
        return res;
    }
};